worker/task_executor.go

```diff
@@ -0,0 +1,164 @@
+package worker
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/ufcg-lsd/arrebol-pb-worker/utils"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+)
+
+const (
+	TaskScriptExecutorFileName  = "task-script-executor.sh"
+	RunTaskScriptCommandPattern = "/bin/bash %s -d -tsf=%s"
+	DefaultWorkerDockerImage    = "ubuntu"
+)
+
+type TaskExecutor struct {
+	Cli client.Client
+	Cid string
+}
+
+func (e *TaskExecutor) Execute(task *Task, containerSpawnWarner chan<- interface{}) error {
+	image := task.Image
+
+	if image == "" {
+		image = DefaultWorkerDockerImage
+	}
+	log.Println("Creating container with image: " + image)
+
+	config := utils.ContainerConfig{
+		Name:   "",
+		Image:  image,
+		Mounts: []mount.Mount{},
+	}
+
+	if err := e.initiate(config); err != nil {
+		return err
+	}
+	if err := e.send(task); err != nil {
+		return err
+	}
+	// It allows the system to know when the container
+	// is ready.
+	containerSpawnWarner <- "spawned"
+	if err := e.run(task.Id); err != nil {
+		task.State = TaskFailed
+		return err
+	}
+	if err := e.stop(); err != nil {
+		return err
+	}
+	task.State = TaskFinished
+	return nil
+}
+
+func (e *TaskExecutor) initiate(config utils.ContainerConfig) error {
+	exists, err := utils.CheckImage(&e.Cli, config.Image)
+	if !exists {
+		if _, err = utils.Pull(&e.Cli, config.Image); err != nil {
+			return err
+		}
+	}
+	cid, err := utils.CreateContainer(&e.Cli, config)
+
+	if err != nil {
+		return err
+	}
+	err = utils.StartContainer(&e.Cli, cid)
+
+	if err != nil {
+		return err
+	}
+
+	err = utils.Exec(&e.Cli, cid, "mkdir /arrebol")
+
+	if err != nil {
+		log.Println("Error on creating /arrebol folder")
+		return err
+	}
+
+	taskScriptExecutorPath := os.Getenv("BIN_PATH") + "/" + TaskScriptExecutorFileName
+
+	err = utils.Copy(&e.Cli, cid, taskScriptExecutorPath, "/arrebol/"+TaskScriptExecutorFileName)
+
+	e.Cid = cid
+	return err
+}
+
+func (e *TaskExecutor) stop() error {
```

Comment:
I think it is misleading to stop and remove the container in a function called stop. why not having two functions?

Reply:
there are 2 functions already in the docker_utils file. Maybe renaming this one? what you think about extinguish? And what do you think about add container in these functions, like initContainer, extinguishContainer ....?

Reply:
let's play simple. First, delete this function, then, in the next rounds we decide what to do

Reply:
ok